File: internal/query/query.go
// MIT License

// Copyright (c) 2026 René-Jean Corneille

// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:

// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.

// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

package query

import (
	"encoding/json"
	"fmt"
	"time"

	"github.com/RonsenbergVI/fraise/internal/config"
	"github.com/RonsenbergVI/fraise/internal/containers"
	"github.com/RonsenbergVI/fraise/internal/graph"
	"github.com/RonsenbergVI/fraise/internal/hash"
	"github.com/RonsenbergVI/fraise/internal/query/parser"
	"github.com/RonsenbergVI/fraise/pkg/logger"
)

type Query[K comparable, P float32 | float64] interface {
	Plan(config *config.ConfigSet) (*Stream[K, P], error)
	GetGraphID() uint8
	hash.Hashable[K, string]
	IsWrite() bool
	SetGraphID(id uint8)
}

type QueryParameters[K comparable] struct {
	Top   int
	Depth int
	Since containers.TimeValue[K]
	Until containers.TimeValue[K]
}

type QueryContext struct {
	GraphID uint8
}

type QueryResult[K comparable, P float32 | float64] struct {
	Count int         `json:"count"`
	Hits  []Hit[K, P] `json:"hits"`

	// Background is the query's background rate ρ₀ — the average seed mass
	// per unit of anchor degree the traversal observed — attached only in
	// explain mode. Together with each hit's contribution breakdown it lets a
	// client recompute every score: S = m + α²·Σ max(0, M_A − m − d_A·ρ₀).
	// omitempty doubles as the mode switch: a plain query never carries it.
	Background P `json:"background,omitempty"`
}

type Hit[K comparable, P float32 | float64] struct {
	Node  *graph.Node[K]
	Score P

	// Contributions is the hit's per-source breakdown, populated only when
	// the stream ran in explain mode; nil otherwise. nil doubles as the
	// serialization switch — a recall hit always has at least one
	// contribution, so nil unambiguously means "not asked for", never
	// "asked for and empty". The entries are wire-shaped (anchor identities
	// already resolved to their values) because resolution needs the graph,
	// which only the commit site holds.
	Contributions []HitContribution[P]
}

// HitContribution is the wire form of one contribution: the source is
// serialized by name, not by its Go constant, because the payload documents
// ranking to clients that never see the enum. A text or vector entry carries
// its raw mass and list position; a graph entry — one per funding anchor —
// carries the anchor's full observed mass, the anchor's value under via, its
// degree, and how many seeds funded it. With the query-level background rate,
// these are exactly the inputs of the scoring fold, so a client can recompute
// the hit's score from its own payload.
type HitContribution[P float32 | float64] struct {
	Source string `json:"source"`
	Score  P      `json:"score"`
	Rank   uint16 `json:"rank"`
	Via    string `json:"via,omitempty"`
	Degree uint32 `json:"degree,omitempty"`
	Count  uint16 `json:"count"`
}

// MarshalJSON flattens the node into the hit so the response carries only the
// value, timestamp and score, with no nested Node object. The contribution
// breakdown appears only when the hit carries one (explain mode), keeping the
// ordinary query response byte-compatible with what it was before explain
// existed.
func (h Hit[K, P]) MarshalJSON() ([]byte, error) {
	node := *h.Node
	source := ""
	if h.Contributions != nil {
		source = node.GetAttributes().Source
	}

	return json.Marshal(struct {
		Value         string               `json:"value"`
		Timestamp     time.Time            `json:"timestamp"`
		Score         P                    `json:"score"`
		Source        string               `json:"source,omitempty"`
		Contributions []HitContribution[P] `json:"contributions,omitempty"`
	}{
		Value:         node.GetValue(),
		Timestamp:     node.GetTimestamp(),
		Score:         h.Score,
		Source:        source,
		Contributions: h.Contributions,
	})
}

// bindVector resolves a vector placeholder to its data, enforcing both that the
// parameter was supplied and that its dimension is within maxDim. A missing
// parameter is ErrMissingParameter; an over-long vector is ErrLimitExceeded —
// both are client errors surfaced as 400, and both are bounded here so a huge
// vector never reaches the index.
func bindVector[P float32 | float64](params map[string][]P, name string, maxDim int) ([]P, error) {
	data, provided := params[name]
	if !provided {
		return nil, fmt.Errorf("%w: $%s", ErrMissingParameter, name)
	}
	if len(data) > maxDim {
		return nil, fmt.Errorf("%w: vector $%s has %d dimensions, max %d", ErrLimitExceeded, name, len(data), maxDim)
	}
	return data, nil
}

// Parse turns a raw query string into an executable Query. Vector arguments are
// passed out-of-band in params, keyed by the placeholder name used in the query
// (e.g. `vec:$v` binds to params["v"]). This keeps the parser lightweight: it
// only records the placeholder, and the real vector is injected here.
//
// Warnings accompany a query that parsed and will run: they flag a reading the
// client may not have meant (see parser.Warning) and must travel to the client
// alongside the results, never attached to the query itself — the plan cache
// substitutes query objects on a hash hit, so state on the query would leak
// between requests.
func Parse[K comparable, P float32 | float64](q string, params map[string][]P, c *config.ConfigSet) (Query[K, P], []parser.Warning, error) {
	cmd, warns, err := parser.Parse[K, P](q)
	if err != nil {
		logger.Debug("Query parsing failed", "query", q, "error", err)
		return nil, nil, fmt.Errorf("%w: %w", ErrParsingFailed, err)
	}

	switch n := cmd.(type) {
	case *parser.RememberCommandNode[P]:
		qo := &Remember[K, P]{
			Value:    n.Value(),
			Entities: n.Entities(),
			Topics:   n.Topics(),
			Source:   n.Source(),
		}
		qo.SetGraphID(n.Selector())

		// Bind the vector placeholder (if any) from the request parameters,
		// rejecting a missing or over-long vector.
		if name, ok := n.VecParam(); ok {
			data, err := bindVector(params, name, c.DB.MaxVectorDimension)
			if err != nil {
				logger.Warn("Rejecting vector parameter for remember", "parameter", name, "error", err)
				return nil, nil, err
			}
			qo.Vector = containers.NewVector[K](data)
		}

		logger.Debug("Parsed remember query", "graph", qo.GetGraphID(), "value", qo.Value)

		return qo, warns, nil

	case *parser.RecallCommandNode[K, P]:
		// Enforce the top/depth ceilings before building the query: a
		// client-supplied result count or walk depth over its ceiling is a
		// client error, rejected here rather than clamped. Only an explicit
		// clause is checked — the configured default is operator-set and trusted,
		// so an unspecified top/depth is never rejected even if the default
		// itself exceeds the ceiling.
		top := n.Top(c.DB.DefaultTop)
		if n.HasTop() && top > c.DB.MaxTop {
			logger.Warn("Rejecting recall over top ceiling", "top", top, "max", c.DB.MaxTop)
			return nil, nil, fmt.Errorf("%w: top:%d exceeds max %d", ErrLimitExceeded, top, c.DB.MaxTop)
		}
		depth := n.Depth(c.DB.DefaultDepth)
		if n.HasDepth() && depth > c.DB.MaxDepth {
			logger.Warn("Rejecting recall over depth ceiling", "depth", depth, "max", c.DB.MaxDepth)
			return nil, nil, fmt.Errorf("%w: depth:%d exceeds max %d", ErrLimitExceeded, depth, c.DB.MaxDepth)
		}

		qo := &Recall[K, P]{
			Keywords: n.Terms(),
			Entities: n.Entities(),
			Topics:   n.Topics(),
			Parameters: QueryParameters[K]{
				Top: top, Depth: depth, Since: n.Since(), Until: n.Until(),
			},
		}
		qo.SetGraphID(n.Selector())

		// Bind the vector placeholder (if any) from the request parameters,
		// rejecting a missing or over-long vector.
		if name, ok := n.VecParam(); ok {
			data, err := bindVector(params, name, c.DB.MaxVectorDimension)
			if err != nil {
				logger.Warn("Rejecting vector parameter for recall", "parameter", name, "error", err)
				return nil, nil, err
			}
			qo.Vector = containers.NewVector[K](data)
		}

		logger.Debug("Parsed recall query", "graph", qo.GetGraphID(), "keywords", len(qo.Keywords))
		return qo, warns, nil

	default:
		logger.Debug("Unsupported query command", "query", q)
		return nil, nil, ErrParsingFailed
	}
}

File: sdk/python/src/fraise_sdk/query.py
# MIT License

# Copyright (c) 2026 René-Jean Corneille

# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:

# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.

# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

"""Builders that turn structured arguments into Fraise query strings.

These are pure functions with no I/O, so they are the natural unit-test seam:
the wire format lives here, and :mod:`fraise_sdk.client` only concerns itself
with transport. The grammar they target:

    remember@<graph> '<value>' [topic:<t>]... [entity:<e>]... [vec:$<name>]
    recall@<graph> <keyword>... [topic:<t>]... [entity:<e>]...
                   [top:<n>] [depth:<n>] [vec:$<name>]

The vector itself never appears in the string — the caller sends it out of band
in the request ``parameters`` map, and only the ``vec:$<name>`` placeholder is
emitted here.
"""

from __future__ import annotations

from collections.abc import Iterable, Sequence

from fraise_sdk.errors import FraiseQueryError

# Name bound to the out-of-band vector in the request parameters. A query carries
# at most one vector, so a single fixed placeholder keeps client and builder in
# lock-step without threading a name through every call.
VECTOR_PARAM = "v"


def _token(kind: str, value: str) -> str:
    """Validate and return a bare grammar token (a keyword, topic or entity).

    The grammar splits on whitespace, so a token may not contain any — a stray
    space would be parsed as two tokens (or fail), silently changing the query.

    Raises:
        FraiseQueryError: if query is not valid
    """
    value = value.strip()
    if not value:
        raise FraiseQueryError(f"{kind} must not be empty")
    if any(ch.isspace() for ch in value):
        raise FraiseQueryError(f"{kind} must not contain whitespace: {value!r}")
    return value


def _clauses(prefix: str, values: Iterable[str] | None) -> list[str]:
    if not values:
        return []
    return [f"{prefix}:{_token(prefix, v)}" for v in values]


def _quote_value(value: str) -> str:
    """Wrap a value (a fact, or a recall's query phrase) in the grammar's quotes.

    Inside a quoted phrase every character is literal and an apostrophe is
    escaped by doubling it (``''``), so any text travels verbatim —
    ``it's blue`` goes over the wire as ``'it''s blue'`` and comes back with
    its apostrophe intact.

    Raises:
        FraiseQueryError: if query is not valid
    """
    if not value.strip():
        raise FraiseQueryError("a quoted value must not be empty")
    escaped = value.replace("'", "''")
    return f"'{escaped}'"


def _selector(graph: int) -> str:
    if not isinstance(graph, int) or isinstance(graph, bool):
        raise FraiseQueryError(f"graph must be an int, got {type(graph).__name__}")
    if graph < 0:
        raise FraiseQueryError(f"graph must be non-negative, got {graph}")
    return f"@{graph}"


def build_remember(
    value: str,
    *,
    graph: int = 0,
    topics: Sequence[str] | None = None,
    entities: Sequence[str] | None = None,
    source: str | None = None,
    with_vector: bool = False,
) -> str:
    """Build a ``remember`` query string that stores ``value`` in ``graph``.

    Set ``with_vector`` when a vector is being sent in the request parameters, so
    the ``vec:$v`` placeholder is appended for the server to bind.
    """
    parts = [f"remember{_selector(graph)}", _quote_value(value)]
    parts += _clauses("topic", topics)
    parts += _clauses("entity", entities)
    if source is not None:
        parts.append(f"source:{_quote_value(source)}")
    if with_vector:
        parts.append(f"vec:${VECTOR_PARAM}")
    return " ".join(parts)


def build_recall(
    keywords: Sequence[str] | None = None,
    *,
    graph: int = 0,
    query: str | None = None,
    topics: Sequence[str] | None = None,
    entities: Sequence[str] | None = None,
    top: int | None = None,
    depth: int | None = None,
    with_vector: bool = False,
) -> str:
    """Build a ``recall`` query string over ``graph``.

    ``query`` is a whole question sent as one quoted phrase term — the grammar
    keeps every character inside the quotes literal, so natural language
    ("What topic has John been blogging about recently?") travels verbatim
    instead of as bare words that would collide with the grammar's reserved
    keywords. ``keywords`` remain individual bare terms and may accompany it.

    A recall needs at least one seed: a query phrase, keywords, a vector, or a
    topic/entity filter. Building one with no seed at all is a programming
    error and is rejected here.

    Raises:
        FraiseQueryError: if query is not valid
    """
    parts = [f"recall{_selector(graph)}"]
    if query is not None:
        parts.append(_quote_value(query))
    parts += [_token("keyword", k) for k in (keywords or [])]
    parts += _clauses("topic", topics)
    parts += _clauses("entity", entities)
    if top is not None:
        if top <= 0:
            raise FraiseQueryError(f"top must be positive, got {top}")
        parts.append(f"top:{top}")
    if depth is not None:
        if depth <= 0:
            raise FraiseQueryError(f"depth must be positive, got {depth}")
        parts.append(f"depth:{depth}")
    if with_vector:
        parts.append(f"vec:${VECTOR_PARAM}")

    if len(parts) == 1:
        raise FraiseQueryError(
            "recall needs at least one seed: keywords, a vector, or a "
            "topic/entity filter"
        )
    return " ".join(parts)
